File: 2017/Day01.hs
module Main where

import Utils (Parser, parseFile)
import Text.Megaparsec.Char (digitChar)
import Text.Megaparsec (many)
import Data.Char (digitToInt)
import Data.List (group, stripPrefix)
import qualified Data.Vector as V

parser :: Parser [Int]
parser = many (digitToInt <$> digitChar)

main :: IO ()
main = do 
  digits <- V.fromList <$> parseFile parser "input/01.txt"

  let score :: (Int -> Int) -> Int
      score partner_of = sum (V.ifilter is_match digits)
        where
          is_match index digit = digits V.! partner_of index == digit

  putStr "Part 1: "
  print $ score $ \index -> (index + 1) `mod` length digits

  putStr "Part 2: "
  print $ score $ \index -> (index + length digits `div` 2) `mod` length digits

File: 2017/day01.rs
use std::fs;

fn captcha<F>(vec: &Vec<u32>, partner_of: F) -> u32
where
    F: Fn(usize) -> u32,
{
    vec.iter()
        .enumerate()
        .filter_map(|(i, digit)| {
            if *digit == partner_of(i) {
                Some(digit)
            } else {
                None
            }
        })
        .sum()
}

fn main() {
    let input = fs::read_to_string("input/01.txt").unwrap();
    let digits: Vec<u32> = input.chars().map_while(|char| char.to_digit(10)).collect();

    let part1_partner_of = |i| digits[(i + 1) % digits.len()];
    let part1: u32 = captcha(&digits, part1_partner_of);

    println!("Part 1: {part1}");

    let part2_get_partner = |i| digits[(i + digits.len() / 2) % digits.len()];
    let part2: u32 = captcha(&digits, part2_get_partner);

    println!("Part 2: {part2}");
}
